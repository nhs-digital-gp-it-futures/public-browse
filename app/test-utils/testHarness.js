import express from 'express';
import nunjucks from 'nunjucks';
import { App } from '../../app';

export const testHarness = () => {
  const app = new App().createApp();
  const router = express.Router();

  return {
    createComponentDummyApp(template, context) {
      const dummyRouter = router.get('/', (req, res) => {
        res.render(template, context);
      });
      app.use(dummyRouter);
      return app;
    },
    createTemplateDummyApp(template, context) {
      const dummyRouter = router.get('/', (req, res) => {
        const viewToTest = nunjucks.renderString(template, context);
        res.send(viewToTest);
      });
      app.use(dummyRouter);
      return app;
    },
  };
};
