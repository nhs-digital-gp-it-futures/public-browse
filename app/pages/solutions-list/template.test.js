import { createTestHarness } from '../../test-utils/testHarness';

const setup = {
  template: {
    path: 'pages/solutions-list/template.njk',
  },
};

describe('solutions list page', () => {
  it('should render the solution list page title', createTestHarness(setup, (harness) => {
    const context = {
      pageTitle: 'some page title',
    };

    harness.request(context, ($) => {
      const solutionListTitle = $('[data-test-id="general-page-title"]');
      expect(solutionListTitle.length).toEqual(1);
      expect(solutionListTitle.text().trim()).toEqual(context.pageTitle);
    });
  }));

  it('should render go back link', createTestHarness(setup, (harness) => {
    const context = {
      pageTitle: 'some page title',
    };

    harness.request(context, ($) => {
      const goBackLink = $('[data-test-id="go-back-link"] a');
      expect(goBackLink.length).toEqual(1);
      expect(goBackLink.text().trim()).toEqual('Go back to previous page');
      expect(goBackLink.attr('href')).toEqual('/solutions');
    });
  }));

  it('should render the solution list page description', createTestHarness(setup, (harness) => {
    const context = {
      pageDescription: 'some page description',
    };

    harness.request(context, ($) => {
      const solutionListTitleSummary = $('div[data-test-id="general-page-description"]');
      expect(solutionListTitleSummary.length).toEqual(1);
      expect(solutionListTitleSummary.text().trim()).toEqual(context.pageDescription);
    });
  }));

  describe('solution cards', () => {
    it('should render 0 cards if no solutions are provided in the context', createTestHarness(setup, (harness) => {
      const context = {
        solutions: [],
      };

      harness.request(context, ($) => {
        const solutionCards = $('div[data-test-id="solution-cards"]').find('[data-test-id="solution-card"]');
        expect(solutionCards.length).toEqual(0);
      });
    }));

    it('should render 1 card if only 1 solution is provided context', createTestHarness(setup, (harness) => {
      const context = {
        solutions: [
          {
            id: '00001',
            name: 'The first solution',
          },
        ],
      };

      harness.request(context, ($) => {
        const solutionCards = $('div[data-test-id="solution-cards"]').find('[data-test-id="solution-card"]');
        expect(solutionCards.length).toEqual(1);
      });
    }));

    it('should render 3 cards if 3 solutions are provided in the context', createTestHarness(setup, (harness) => {
      const context = {
        solutions: [
          {
            id: '00001',
            name: 'The first solution',
          },
          {
            id: '00002',
            name: 'The second solution',
          },
          {
            id: '00003',
            name: 'The third solution',
          },
        ],
      };

      harness.request(context, ($) => {
        const solutionCards = $('div[data-test-id="solution-cards"]').find('[data-test-id="solution-card"]');
        expect(solutionCards.length).toEqual(3);
      });
    }));
  });
});
