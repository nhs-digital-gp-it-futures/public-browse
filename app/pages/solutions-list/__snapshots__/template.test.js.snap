// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`solutions list page should render the content for capabilities-selector filter type 1`] = `
"<main class=\\"nhsuk-main-wrapper\\" id=\\"maincontent\\" data-test-id=\\"main-content\\">

    
  <div class=\\"nhsuk-grid-row nhsuk-width-container\\" data-test-id=\\"solutions-list-body\\">

    <div class=\\"nhsuk-u-padding-top-5\\" data-test-id=\\"go-back-link\\">
      <div class=\\"nhsuk-back-link\\">
  <a class=\\"nhsuk-back-link__link\\" href=\\"#\\">
  <svg class=\\"nhsuk-icon nhsuk-icon__chevron-left\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\">
    <path d=\\"M8.5 12c0-.3.1-.5.3-.7l5-5c.4-.4 1-.4 1.4 0s.4 1 0 1.4L10.9 12l4.3 4.3c.4.4.4 1 0 1.4s-1 .4-1.4 0l-5-5c-.2-.2-.3-.4-.3-.7z\\"></path>
  </svg>
  Go back to previous page</a>
</div>

    </div>

    
  <div class=\\"nhsuk-grid-row\\" data-test-id=\\"general-description\\">
    <div class=\\"nhsuk-grid-column-two-thirds\\">
      
        <h1 class=\\"nhsuk-u-margin-top-3 nhsuk-u-margin-bottom-3\\">Catalogue Solution – results</h1>
      
      
        <p class=\\"nhsuk-heading-l nhsuk-u-font-size-22\\">These are the Catalogue Solutions that match your search criteria.</p>
      
    </div>
  </div>


    
      <div class=\\"nhsuk-grid-row\\">
        <div data-test-id=\\"compare-solutions-description\\" class=\\"nhsuk-u-font-size-22 nhsuk-grid-column-two-thirds\\">You can use our compare document to help find what you’re looking for. It gives an overview of all the Catalogue Solutions available and lets you filter them by the Capabilities they offer.</div>
          <div data-test-id=\\"compare-button\\" class=\\"nhsuk-grid-column-two-thirds nhsuk-u-margin-top-5\\">
            


  
    
  




<a class=\\"nhsuk-button nhsuk-button--secondary\\" href=\\"/solutions/compare/document\\" draggable=\\"false\\">
  Compare Solutions (.xlsx)
</a>



          </div>
        </div>
      </div>
    

    <div data-test-id=\\"solution-cards\\" class=\\"nhsuk-u-margin-bottom-7 nhsuk-u-margin-top-7\\">
      
      <div class=\\"bc-t-border-top--blue\\">
    </div>
  </div>


  </main>"
`;

exports[`solutions list page should render the content for foundation filter type 1`] = `
"<main class=\\"nhsuk-main-wrapper\\" id=\\"maincontent\\" data-test-id=\\"main-content\\">

    
  <div class=\\"nhsuk-grid-row nhsuk-width-container\\" data-test-id=\\"solutions-list-body\\">

    <div class=\\"nhsuk-u-padding-top-5\\" data-test-id=\\"go-back-link\\">
      <div class=\\"nhsuk-back-link\\">
  <a class=\\"nhsuk-back-link__link\\" href=\\"#\\">
  <svg class=\\"nhsuk-icon nhsuk-icon__chevron-left\\" xmlns=\\"http://www.w3.org/2000/svg\\" viewBox=\\"0 0 24 24\\" aria-hidden=\\"true\\">
    <path d=\\"M8.5 12c0-.3.1-.5.3-.7l5-5c.4-.4 1-.4 1.4 0s.4 1 0 1.4L10.9 12l4.3 4.3c.4.4.4 1 0 1.4s-1 .4-1.4 0l-5-5c-.2-.2-.3-.4-.3-.7z\\"></path>
  </svg>
  Go back to previous page</a>
</div>

    </div>

    
  <div class=\\"nhsuk-grid-row\\" data-test-id=\\"general-description\\">
    <div class=\\"nhsuk-grid-column-two-thirds\\">
      
        <h1 class=\\"nhsuk-u-margin-top-3 nhsuk-u-margin-bottom-3\\">Foundation Solution Sets – results</h1>
      
      
        <p class=\\"nhsuk-heading-l nhsuk-u-font-size-22\\">These Catalogue Solutions meet the 6 Foundation Capabilities that are the minimum requirement to enable a GP practice to operate.</p>
      
    </div>
  </div>


    

    <div data-test-id=\\"solution-cards\\" class=\\"nhsuk-u-margin-bottom-7 nhsuk-u-margin-top-7\\">
      
      <div class=\\"bc-t-border-top--blue\\">
    </div>
  </div>


  </div></main>"
`;
